File: 401-500/426-convert-binary-search-tree-to-sorted-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};
*/
class Solution {
    Node* first = nullptr;
    Node* prev = nullptr;
public:
    Node* treeToDoublyList(Node* root) {
        if (root == nullptr) return nullptr;
        helper(root);
        first->left = prev;
        prev->right = first;

        return first;
    }

    void helper(Node *root) {
        if (root == nullptr) return;
        helper(root->left);

        if (first == nullptr) {
            first = root;
            prev = first;
        }
        else {
            prev->right = root;
            root->left = prev;
            prev = root;
        }

        helper(root->right);
    }
};

File: python version/426-convert-binary-search-tree-to-sorted-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""
class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if not root:
            return None
        self.prev = None

        def helper(node):
            if not node:
                return
            helper(node.left)
            if self.prev:
                self.prev.right = node
                node.left = self.prev
            else:
                self.head = node
            self.prev = node
            helper(node.right)
        
        helper(root)

        self.head.left = self.prev
        self.prev.right = self.head

        return self.head

            
        